source/enum_set.h

```diff
@@ -12,195 +12,280 @@
-      if (overflow_->find(item) != overflow_->end()) return true;
+  // Returns the `enum` stored in `bucket` at `offset`.
+  // `offset` is the bit-offset in the bucket storage.
+  static constexpr inline T GetValueFromBucket(const Bucket& bucket,
+                                               ElementType offset) {
+    return static_cast<T>(static_cast<ElementType>(bucket.start) + offset);
+  }
+
+  // For a given enum `value`, finds the bucket index that could contain this
+  // value. If no such bucket is found, the index at which the new bucket should
+  // be inserted is returned.
+  size_t FindBucketForValue(T value) const {
+    // Set is empty, insert at 0.
+    if (buckets_.size() == 0) {
+      return 0;
     }
 
-    return false;
-  }
+    size_t index =
+        std::min(buckets_.size() - 1, ComputeTheoreticalBucketIndex(value));
+    const T needle = ComputeBucketStart(value);
 
- private:
-  // Adds the given enum value (as a 32-bit word) to the set.  This has no
-  // effect if the enum value is already in the set.
-  void AddWord(uint32_t word) {
-    if (auto new_bits = AsMask(word)) {
-      mask_ |= new_bits;
-    } else {
-      Overflow().insert(word);
+    const T bucket_start = buckets_[index].start;
+    // Computed index is the correct one.
+    if (bucket_start == needle) {
+      return index;
     }
-  }
 
-  // Removes the given enum value (as a 32-bit word) from the set.  This has no
-  // effect if the enum value is not in the set.
-  void RemoveWord(uint32_t word) {
-    if (auto new_bits = AsMask(word)) {
-      mask_ &= ~new_bits;
-    } else {
-      auto itr = Overflow().find(word);
-      if (itr != Overflow().end()) Overflow().erase(itr);
+    // Bucket contains smaller values. Linear scan right.
+    if (bucket_start < needle) {
+      for (index += 1;
+           index < buckets_.size() && buckets_[index].start < needle; index++) {
+      }
+      return index;
     }
-  }
 
-  // Returns true if the enum represented as a 32-bit word is in the set.
-  bool ContainsWord(uint32_t word) const {
-    // We shouldn't call Overflow() since this is a const method.
-    if (auto bits = AsMask(word)) {
-      return (mask_ & bits) != 0;
-    } else if (auto overflow = overflow_.get()) {
-      return overflow->find(word) != overflow->end();
+    // Bucket contains larger values, insert front.
+    if (index == 0) {
+      return index;
     }
-    // The word is large, but the set doesn't have large members, so
-    // it doesn't have an overflow set.
-    return false;
-  }
 
-  // Returns the enum value as a uint32_t.
-  uint32_t ToWord(EnumType value) const {
-    static_assert(sizeof(EnumType) <= sizeof(uint32_t),
-                  "EnumType must statically castable to uint32_t");
-    return static_cast<uint32_t>(value);
+    for (index -= 1; index > 0 && buckets_[index].start > needle; index--) {
+    }
+    return buckets_[index].start >= needle ? index : index + 1;
```

Comment:
I'm not sure how you want to use lower/upper bound:

* if the idea is to replace the whole logic with a single std::lower_bound():
Sure, that's definitely easier, 1 single call to lower_bound, makes it easier to follow.
But now we don't use the bucket indexing to search with an heuristic, we do a linear scan for each insertion. (Roughly we pack enums into a sorted vector).

Not having optimized code for clarity-sake is an argument, but it goes against your other comment suggesting not to to a linear scan, but to do a merge-like algorithm for the `HasAnyOf` (as in trade code simplicity for optimization there).

* if the idea is to replace the for-loops with std::lower_bound/std::upper_bound:

 That's also a viable option, as you don't have to read the loop to understand what it does, since upper_bound/lower_bound is supposed to be known and is well documented.
But implementing it, I'm not convinced it reduces the off-by-one risk:
 - lower_bound does replace the first for loop in a clean way.
 - left scan with upper_bound requires reverse iterators, meaning we have a +1 showing up.
 - then `distance(begin, reverse(it))` means doing some `size() - distance(it)` operations, which do feel more dangerous than the for look implemented here.

Reply:
(checking actual impact of using the packed vector with a dumb lower_bound, but not doing a left/right scan from a predicted index)

Reply:
Replacing the whole heuristic logic with a simple `lower_bound()` (which makes the code way simpler) means a 10% hit on performance for sorted insertions, 40% hit when doing random operations (full benchmark, +400% on the benchmark on small enums only..).